hathor/manager.py

```diff
@@ -482,30 +494,43 @@ def on_new_tx(self, tx: BaseTransaction, *, conn: Optional[HathorProtocol] = Non
                 self.log.debug('on_new_tx(): Already have transaction {}'.format(tx.hash.hex()))
                 return False
 
-        try:
-            assert self.validate_new_tx(tx) is True
-        except (InvalidNewTransaction, TxValidationError) as e:
-            # Discard invalid Transaction/block.
-            self.log.debug('Transaction/Block discarded', tx=tx, exc=e)
-            if not fails_silently:
-                raise
-            return False
+        if self.state != self.NodeState.INITIALIZING or not self._fast_init:
+            try:
+                assert self.validate_new_tx(tx) is True
+            except (InvalidNewTransaction, TxValidationError) as e:
+                # Discard invalid Transaction/block.
+                self.log.debug('Transaction/Block discarded', tx=tx, exc=e)
+                if not fails_silently:
+                    raise
+                return False
 
         if self.state != self.NodeState.INITIALIZING:
             self.tx_storage.save_transaction(tx)
         else:
-            tx.reset_metadata()
+            if not self._fast_init:
+                tx.reset_metadata()
+
             self.tx_storage._add_to_cache(tx)
 
-        try:
-            tx.update_initial_metadata()
-            self.consensus_algorithm.update(tx)
-        except Exception:
-            pretty_json = json.dumps(tx.to_json(), indent=4)
-            self.log.error('An unexpected error occurred when processing {tx.hash_hex}\n'
-                           '{pretty_json}', tx=tx, pretty_json=pretty_json)
-            self.tx_storage.remove_transaction(tx)
-            raise
+            if self._fast_init:
```

Comment:
I think we could merge the code a bit:
```
self.tx_storage._add_to_cache(tx)
if not self._fast_init:
    tx.reset_metadata()
else:
    # the code below
```

From what I've seen, there's no harm in adding to the index before resetting the metadata.

Reply:
Yeah, it's the same. Did the merge you suggested.